### PR TITLE
Extrusions of subsequent sections

### PR DESCRIPTION
**Pull request by morettid**

### CODE REVIEW

File: benchmark_tests/test_moving_load_on_3_sections_3d/test_moving_load_on_3_sections_3d.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    # Specify the coordinates for the shapes to extrude: x, y (z is ignored if defined, overwritten by sections)
+
+    embankment_coordinates = [(0.0, 0.0, 0.0), (3.0, 0.0, 0.0), (1.5, 1.0, 0.0), (0, 1.0, 0.0)]
+    bridge_coordinates = [(0.0, 0.8, 0.0), (1.0, 0.8, 0.0), (1.0, 1.0, 0.0), (0., 1.0, 0.0)]
+
+    # Create the soil layer
+    model.add_soil_layer_by_coordinates(embankment_coordinates, material_soil, "embankment1", "Section 1")
+    model.add_soil_layer_by_coordinates(bridge_coordinates, material_bridge, "bridge", "Section 2")
+    model.add_soil_layer_by_coordinates(embankment_coordinates, material_soil, "embankment2", "Section 3")
+    model.show_geometry(show_surface_ids=True)
```
[the remark this answers is not in the document]
Comment (aronnoordam) [reply benchmark_tests/test_moving_load_on_3_sections_3d/test_moving_load_on_3_sections_3d.py:68]: else it is annoying when running the benchmark tests

File: benchmark_tests/test_moving_load_on_3_sections_3d/test_moving_load_on_3_sections_3d.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+from stem.model import Model
+from stem.output import NodalOutput, VtkOutputParameters
+from stem.soil_material import OnePhaseSoil, LinearElasticSoil, SoilMaterial, SaturatedBelowPhreaticLevelLaw
+from stem.solver import AnalysisType, SolutionType, TimeIntegration, DisplacementConvergenceCriteria, \
+    StressInitialisationType, SolverSettings, Problem
+from stem.stem import Stem
+
+IS_LINUX = sys.platform == "linux"
+
+@pytest.mark.skipif(IS_LINUX, reason="calculate_moving_load_on_3_sections_3d: test not implemented for linux system yet."
```
Comment (aronnoordam) [inline benchmark_tests/test_moving_load_on_3_sections_3d/test_moving_load_on_3_sections_3d.py:19]: can you make an issue for this

File: stem/model.py [hunk 3]
```diff
@@ -210,35 +213,122 @@ def add_all_layers_from_geo_file(self, geo_file_name: str, body_names: Sequence[
...
+                if group_name not in self.groups.keys():
+                    raise ValueError(f"Non-existent group specified `{group_name}`.")
+
+                # retrieve end and start z coordinate for extruding the section
+                extrusion_parameters = self.groups[group_name]["extrusion_parameters"]
+                # extrusion length is computed as the difference between end and start z coordinate.
+                direction_vector = extrusion_parameters["direction_vector"]
+                extrusion_vector: List[float] = [dv * extrusion_parameters["length"] for dv in direction_vector]
+                gmsh_input[name]["extrusion_length"] = extrusion_vector
+                initial_extrusion_coordinate = extrusion_parameters["start_coord"][OUT_OF_PLANE_AXIS_2D]
```
[the remark this answers is not in the document]
Comment (morettid) [reply stem/model.py:318]: In this case then the user has to specify the Z coordinate as well in the coordinates of the soil layer.

File: stem/model.py [hunk 4]
```diff
@@ -210,35 +211,123 @@ def add_all_layers_from_geo_file(self, geo_file_name: str, body_names: Sequence[
...
+                direction_vector = extrusion_parameters["direction_vector"]
+                norm = np.linalg.norm(direction_vector)
+                extrusion_vector: List[float] = [dv * extrusion_parameters["length"] / norm for dv in direction_vector]
+                gmsh_input[name]["extrusion_length"] = extrusion_vector
+
+                reference_point_group = extrusion_parameters["reference_coordinate"]
+
+                if not Utils.is_point_coplanar_to_polygon(reference_point_group, coordinates):
+                    raise ValueError("The polygon specified for the soil layer doesn't contain the reference point for "
+                                     "the group.")
```
Comment (aronnoordam) [inline stem/model.py:322]:
```suggestion
                    raise ValueError(f"The reference coordinate of group: {group_name}, does not lay on the same plane as soil layer: {name}))
```

**Review by aronnoordam — APPROVED** [review]

**Pull request merged**